qiskit/extensions/hamiltonian_gate.py

```diff
@@ -0,0 +1,124 @@
+
+from .unitary import UnitaryGate
+
+
+class HamiltonianGate(Gate):
+    """Class for representing evolution by a Hermitian Hamiltonian operator as a gate. This gate
+    resolves to a UnitaryGate, which can be decomposed into basis gates if it is 2 qubits or
+    less, or simulated directly in Aer for more qubits. """
+
+    def __init__(self, data, time, label=None):
+        """Create a gate from a hamiltonian operator and evolution time parameter t
+
+        Args:
+            data (matrix or Operator): unitary operator.
+            time (float or complex): time evolution parameter.
+            label (str): unitary name for backend [Default: None].
+
+        Raises:
+            ExtensionError: if input data is not an N-qubit unitary operator.
+        """
+        if hasattr(data, 'to_matrix'):
+            # If input is Gate subclass or some other class object that has
+            # a to_matrix method this will call that method.
+            data = data.to_matrix()
+        elif hasattr(data, 'to_operator'):
+            # If input is a BaseOperator subclass this attempts to convert
+            # the object to an Operator so that we can extract the underlying
+            # numpy matrix from `Operator.data`.
+            data = data.to_operator().data
+        # Convert to numpy array in case not already an array
+        data = numpy.array(data, dtype=complex)
+        # Check input is unitary
+        if not is_hermitian_matrix(data):
+            raise ExtensionError("Input matrix is not Hermitian.")
+        # Check input is N-qubit matrix
+        input_dim, output_dim = data.shape
+        num_qubits = int(numpy.log2(input_dim))
+        if input_dim != output_dim or 2**num_qubits != input_dim:
+            raise ExtensionError(
+                "Input matrix is not an N-qubit operator.")
+
+        # Store instruction params
+        super().__init__('hamiltonian', num_qubits, [data, time], label=label)
+
+    def __eq__(self, other):
+        if not isinstance(other, HamiltonianGate):
+            return False
+        if self.label != other.label:
+            return False
+        # Should we match unitaries as equal if they are equal
+        # up to global phase?
+        operators_eq = matrix_equal(self.params[0], other.params[0], ignore_phase=True)
+        times_eq = self.params[1] == other.params[1]
+        return operators_eq and times_eq
+
+    def to_matrix(self):
+        """Return matrix for the unitary."""
+        unbound_t = False
+        if isinstance(self.params[1], ParameterExpression):
+            unbound_t = len(self.params[1].parameters) > 0
+        if isinstance(self.params[1], Number) or not unbound_t:
+            # pylint: disable=no-member
+            return scipy.linalg.expm(1j * self.params[0] * self.params[1])
+        else:
+            raise NotImplementedError("Unable to generate Unitary matrix for "
+                                      "unbound t parameter {}".format(self.params[1]))
```

Comment:
```suggestion
        try:
            # pylint: disable=no-member
            return scipy.linalg.expm(1j * self.params[0] * float(self.params[1]))
         except TypeError:
            raise TypeError("Unable to generate Unitary matrix for "
                            "unbound t parameter {}".format(self.params[1]))
```
Calling `float` on the `ParameterExpression` will raise a `TypeError` if it is unbound, and will give you  float you can multiple by `1j` (so you won't need the parameter changes).